### DISCUSSION TITLE
TUnit 0.3.0 has broken my test code

### DISCUSSION THREAD

**Discussion by Patrick8639** [thomhurst/TUnit#1242] [Q&A] [answered]
Hello everybody,

I have converted all my test code from `xUnit` to `TUnit` and it works great under version `0.2.212`.
Updating TUnit to version `3.0.x` makes my test code not compiling, all the errors being when using the `MethodDataSource` attribute.

Here is an example of what doesn't work:

The base class for all my JSON values is defined as:
```cs
public abstract class JsonValue
```

To test casting to a `JsonBooleanValue`, I have the following test:
```cs
  public static IEnumerable <JsonValue> InvalidCastValues () =>
  [
    new JsonArrayValue    (null),
    new JsonDateValue     (null, Date.Today ()),
    new JsonDateTimeValue (null, DateTime.Now),
    new JsonDoubleValue   (null, 0.0),
    new JsonInt32Value    (null, 0),
    new JsonInt64Value    (null, 0),
    new JsonObjectValue   (new JsonInt32Value ("a", 0))
  ];



  [Test]
  [MethodDataSource (nameof (InvalidCastValues))]
  public async Task Cast_FromInvalidValue (
    JsonValue json
  ) =>
    await Assert.That (() => _ = (Boolean) json).ThrowsExactly <ArgumentException> ();
```

This code compiles well with `TUnit 0.2.212`, but, under `TUnit 0.3.xxx`I have the following errors:
> Error CS1503 : Argument 1: cannot convert from 'System.Collections.Generic.List<OrdinaSoft.Json.JsonValue>' to 'OrdinaSoft.Json.JsonValue'
> Warning TUnit0046 : Return a `Func<T>` rather than a `<T>`.

I've tried to change the data source method to the following, but I still have compilation errors:
```cs
  public static IEnumerable <Func <JsonValue>> InvalidCastValues () =>
  [
    static () => new JsonArrayValue    (null),
    static () => new JsonDateValue     (null, Date.Today ()),
    static () => new JsonDateTimeValue (null, DateTime.Now),
    static () => new JsonDoubleValue   (null, 0.0),
    static () => new JsonInt32Value    (null, 0),
    static () => new JsonInt64Value    (null, 0),
    static () => new JsonObjectValue   (new JsonInt32Value ("a", 0))
  ];
```

> In the `MethodDataSoruce` atrribute - Error TUnit0001 : Attribute argument types 'System.Collections.Generic.List<OrdinaSoft.Json.JsonValue>' don't match method parameter types 'OrdinaSoft.Json.JsonValue'

And also errors in the test method.

I've also tried to use `yield return` in my data source, but with the same errors.

I'm currently out of ideas to make this work. Does somebody knows what's wrong?

**Comment by thomhurst** [comment]
Sounds like a bug in a recent refactor @Patrick8639.

Could you produce me a minimal repro? I've tried myself with a simple example but it's not giving me any errors:

```csharp
using TUnit.Assertions;
using TUnit.Assertions.Extensions;

namespace TUnit.TestProject;

public class Repro
{    
    [Test]
    [MethodDataSource(nameof(MethodWithBaseReturn))]
    public void DataSource_WithBaseReturn(BaseValue value)
    {
    }

    public static IEnumerable<Func<BaseValue>> MethodWithBaseReturn() =>
    [
        () => new ConcreteValue(),
        () => new ConcreteValue2()
    ];
    
    public abstract class BaseValue;

    public class ConcreteValue : BaseValue;
    public class ConcreteValue2 : BaseValue;
}
```

**Comment by thomhurst** [comment]
Your error message says you're returning a `System.Collections.Generic.List`, but your code says `IEnumerable` - Did you change it?

**Comment by Patrick8639** [comment]
@thomhurst If you have a private place, I can give you the source code of the classes and the tests.

**Comment by Patrick8639** [comment]
@thomhurst 
> Your error message says you're returning a System.Collections.Generic.List, but your code says IEnumerable - Did you change it?

Sorry for this: I have multiple tests like that (one for each different data type) and forget to convert one of them to `IEnumerable`. After changing it, the tests compile but I still have the following warning:
> Warning TUnit0046 : Return a Func<T> rather than a <T>.

If I change the `InvalidCastValues` to return `IEnumerable <Func <JsonValue>>`, I have the error in the `MethodDataSource` attribute.
But the tests run without any error!

**Comment by Patrick8639** [comment]
So the only problem that remains is the following error displayed in `Rider`:
![image](https://github.com/user-attachments/assets/8ce40992-526a-459e-a697-78f0e08bd500)

**Reply by thomhurst** [reply, marked as answer]
Can you try v0.3.29?

**Comment by Patrick8639** [comment]
> Can you try v0.3.29?

It works perfectly with version 0.3.29.
Thanks for the good work.
